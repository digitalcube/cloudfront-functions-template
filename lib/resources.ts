import { Function } from 'cff-tools';
import { join } from 'path';

const stage = process.env.STAGE || 'development'
export const ViewerRequestFunction = new Function({
  name: 'ViewerRequestFunction-' + stage,
  runtime: 'cloudfront-js-1.0'
}, {
    functionFilePath: join(__dirname, '../functions/viewer_reqeust.js')
})

export const ViewerResponseFunction = new Function({
  name: 'ViewerResponseFunction-' + stage,
  runtime: 'cloudfront-js-1.0'
}, {
    functionFilePath: join(__dirname, '../functions/viewer_response.js')
})
