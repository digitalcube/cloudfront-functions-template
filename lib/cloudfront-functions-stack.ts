import { CfnFunction } from '@aws-cdk/aws-cloudfront';
import * as cdk from '@aws-cdk/core';
import { ViewerRequestFunction, ViewerResponseFunction } from './resources'

export class CloudfrontFunctionsStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    const stage = process.env.STAGE || 'development'

    // The code that defines your stack goes here
    new CfnFunction(this, 'ViewerRequestFunction' + stage, {
      functionCode: ViewerRequestFunction.getFunctionCode().toString(),
      autoPublish: false,
      functionConfig: {
        comment: "empty function (boiler plate)",
        runtime: ViewerRequestFunction.runtime
      },
      name: ViewerRequestFunction.name
    })

    new CfnFunction(this, 'ViewerResponseFunction' + stage, {
      name: ViewerResponseFunction.name,
      autoPublish: false,
      functionCode: ViewerResponseFunction.getFunctionCode().toString(),
      functionConfig: {
        comment: "Add easing security header and set redirection of s3 object uuid removing",
        runtime: ViewerResponseFunction.runtime
      }
    })
  }
}
