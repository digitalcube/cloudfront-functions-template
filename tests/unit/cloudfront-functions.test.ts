import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as cdk from '@aws-cdk/core';
import * as CloudfrontFunctions from '../../lib/cloudfront-functions-stack';

describe('unit test for the stack', () => {
  let app: cdk.App
  let stack: cdk.Stack
  beforeEach(() => {
    app = new cdk.App();
    stack = new CloudfrontFunctions.CloudfrontFunctionsStack(app, 'MyTestStack');
  })
  test('Should match snapshot', () => {
      expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  });
  it('should have CloudFront Function resources', () => {
    expect(stack).toHaveResource( "AWS::CloudFront::Function")
  })
  it('should disabled auto publish attributes', () => {
    expect(stack).toHaveResource( "AWS::CloudFront::Function", {
      AutoPublish: false,
    })
  })
  it('should not contain module.export on the functionCode', () => {
    const resources = Object.values(SynthUtils.toCloudFormation(stack).Resources)
      .filter((resource: any) => resource.Type === "AWS::CloudFront::Function" )
    resources.forEach((resource: any) => {
      expect(resource.Properties.FunctionCode).not.toContain('module.exports')
    })
  })
  it('should not contain const on the functionCode', () => {
    const resources = Object.values(SynthUtils.toCloudFormation(stack).Resources)
      .filter((resource: any) => resource.Type === "AWS::CloudFront::Function" )
    resources.forEach((resource: any) => {
      expect(resource.Properties.FunctionCode).not.toContain('const')
    })
  })
  it('should not contain let on the functionCode', () => {
    const resources = Object.values(SynthUtils.toCloudFormation(stack).Resources)
      .filter((resource: any) => resource.Type === "AWS::CloudFront::Function" )
    resources.forEach((resource: any) => {
      expect(resource.Properties.FunctionCode).not.toContain('let')
    })
  })
})