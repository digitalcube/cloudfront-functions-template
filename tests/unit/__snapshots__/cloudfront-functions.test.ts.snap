// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for the stack Should match snapshot 1`] = `
Object {
  "Resources": Object {
    "ViewerRequestFunctiondevelopment": Object {
      "Properties": Object {
        "AutoPublish": false,
        "FunctionCode": "function handler(event) {
    var request = event.request;
    return request;
}",
        "FunctionConfig": Object {
          "Comment": "empty function (boiler plate)",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": "ViewerRequestFunction-development",
      },
      "Type": "AWS::CloudFront::Function",
    },
    "ViewerResponseFunctiondevelopment": Object {
      "Properties": Object {
        "AutoPublish": false,
        "FunctionCode": "function handler(event) {
    var response = event.response;    
    // Set HTTP security headers
    response.headers['strict-transport-security'] = { value: 'max-age=63072000'}; 
    response.headers['x-content-type-options'] = { value: 'nosniff'}; 
    response.headers['x-frame-options'] = {value: 'SAMEORIGIN'}; 
    response.headers['x-xss-protection'] = {value: '1; mode=block'}; 
    return response;
  }",
        "FunctionConfig": Object {
          "Comment": "Add easing security header and set redirection of s3 object uuid removing",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": "ViewerResponseFunction-development",
      },
      "Type": "AWS::CloudFront::Function",
    },
  },
}
`;
